File: algoritmos/java/CASE.Algoritimos/src/main/MergeSort.java
package main;

public class MergeSort
{
	private void merge(int[] array, int l, int m, int r)
	{
		int n1 = m - l + 1;
		int n2 = r - m;

		int[] esquerda = new int [n1];
		int[] direita = new int [n2];

		for (int a = 0; a < n1; ++a)
        {
            esquerda[a] = array[l + a];
        }

        for (int b = 0; b < n2; ++b)
        {
            direita[b] = array[m + 1 + b];
        }

		int i = 0, j = 0;

		int k = l;
		while (i < n1 && j < n2)
		{
			if (esquerda[i] <= direita[j])
			{
				array[k] = esquerda[i];
				i++;
			}
			else
			{
				array[k] = direita[j];
				j++;
			}
			k++;
		}

		while (i < n1)
		{
			array[k] = esquerda[i];
			i++;
			k++;
		}

		while (j < n2)
		{
			array[k] = direita[j];
			j++;
			k++;
		}
	}

	public void sort(int[] array, int l, int r)
	{
		if (l < r)
		{
			int m = (l+r)/2;
			sort(array, l, m);
			sort(array , m+1, r);
			merge(array, l, m, r);
		}
	}
}

File: algoritmos/python/CASE.Algoritmos/MergeSort.py
class MergeSort:
    def merge(self, array, l, m, r):
        n1 = (m - l + 1)
        n2 = r - m

        esquerda = [0] * (n1)
        direita = [0] * (n2)

        for i in range(0, n1):
            esquerda[i] = array[l + i]

        for j in range(0, n2):
            direita[j] = array[m + 1 + j]

        i = 0
        j = 0
        k = l
        while i < n1 and j < n2:
            if esquerda[i] <= direita[j]:
                array[k] = esquerda[i]
                i += 1
            else:
                array[k] = direita[j]
                j += 1
            k += 1

        while i < n1:
            array[k] = esquerda[i]
            i += 1
            k += 1

        while j < n2:
            array[k] = direita[j]
            j += 1
            k += 1


    def sort(self, array, l, r):
        if l < r:
            m = (l + (r - 1)) // 2
            self.sort(array, l, m)
            self.sort(array, m + 1, r)
            self.merge(array, l, m, r)

